### ISSUE TITLE
Entrenar

### ISSUE THREAD

**Issue by amauryma55** [amauryma55/python#3]
# Separación de X e y
columna_objetivo = 'var_res'
X = df_aldia_007.drop(columns=[columna_objetivo])
y = df_aldia_007[columna_objetivo]

# Train-test split estratificado
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

# Preprocesador SOLO para variables categóricas
preprocesador = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(handle_unknown='ignore'), X.columns)  # Todas las columnas
    ],
    remainder='drop'
)

# Función de evaluación
def evaluar_modelo(nombre, y_test, y_pred, y_prob):
    print(f"\n=== {nombre} ===")
    print(confusion_matrix(y_test, y_pred))
    print(classification_report(y_test, y_pred))
    print(f'ROC-AUC: {roc_auc_score(y_test, y_prob):.4f}')

# Cross-validation estratificada
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

RL

pipe_log = Pipeline([
    ('preprocesador', preprocesador),
    ('modelo', LogisticRegression(class_weight='balanced', max_iter=1000, random_state=42))
])

# Grid de búsqueda de hiperparámetros
param_grid_log = {
    'modelo__C': [0.01, 0.1, 1, 10, 100],
    'modelo__penalty': ['l2'],
    'modelo__solver': ['lbfgs'],
}

# GridSearchCV con AUC como métrica principal
grid_log = GridSearchCV(pipe_log, param_grid_log, cv=cv, scoring='roc_auc', n_jobs=-1)
grid_log.fit(X_train, y_train)

# Predicciones
y_pred_log = grid_log.predict(X_test)
y_prob_log = grid_log.predict_proba(X_test)[:, 1]

# Evaluación
evaluar_modelo("Regresión Logística", y_test, y_pred_log, y_prob_log)
print("Mejores parámetros regresión logística:", grid_log.best_params_)

RF

pipe_rf = Pipeline([
    ('preprocesador', preprocesador),
    ('modelo', RandomForestClassifier(class_weight='balanced', random_state=42))
])

param_grid_rf = {
    'modelo__n_estimators': [100, 200, 300],
    'modelo__max_depth': [None, 5, 10, 20],
    'modelo__min_samples_split': [2, 5, 10],
}

grid_rf = GridSearchCV(pipe_rf, param_grid_rf, cv=cv, scoring='roc_auc', n_jobs=-1)
grid_rf.fit(X_train, y_train)

y_pred_rf = grid_rf.predict(X_test)
y_prob_rf = grid_rf.predict_proba(X_test)[:, 1]

evaluar_modelo("Random Forest", y_test, y_pred_rf, y_prob_rf)
print("Mejores parámetros Random Forest:", grid_rf.best_params_)

XGB

from xgboost import XGBClassifier

# Ajuste de clase desbalanceada
n_0 = np.sum(y_train == 0)
n_1 = np.sum(y_train == 1)
scale_pos_weight = n_0 / n_1

pipe_xgb = Pipeline([
    ('preprocesador', preprocesador),
    ('modelo', XGBClassifier(
        scale_pos_weight=scale_pos_weight,
        use_label_encoder=False,
        eval_metric='logloss',
        random_state=42
    ))
])

param_grid_xgb = {
    'modelo__n_estimators': [100, 200, 300],
    'modelo__max_depth': [3, 5, 7, 10],
    'modelo__learning_rate': [0.01, 0.1, 0.2],
    'modelo__subsample': [0.8, 1.0]
}

grid_xgb = GridSearchCV(pipe_xgb, param_grid_xgb, cv=cv, scoring='roc_auc', n_jobs=-1)
grid_xgb.fit(X_train, y_train)

y_pred_xgb = grid_xgb.predict(X_test)
y_prob_xgb = grid_xgb.predict_proba(X_test)[:, 1]

evaluar_modelo("XGBoost", y_test, y_pred_xgb, y_prob_xgb)
print("Mejores parámetros XGBoost:", grid_xgb.best_params_)
